### PR TITLE
LIME-2053 - Add functionality to query cloud watch log groups within …

### PR DESCRIPTION
**Pull request by ChrisBates1**
Added in PiiTermsCollector utility with method to build a set of PII values from the TestDataCreator i.e. first names, last names, licence numbers, full DOBs, postcodes). 
Added TestRunContext utility class with a static suiteStartTime set once when the scenarios first execute (through a @BeforeAll)

Added PiiLogScanStepDefs with @AfterHook, this skips the check is the AWS_STACK_NAME is not set, It then checks over each available log group and ECS Log Group, and queries CloudWatch for each PII terms, and asserts that non are found

If any PII Data match is found within the available logs, the test will fail and the PII exposed will be displayed in the logs. 

### Issue tracking

- [LIME-2053](https://govukverify.atlassian.net/browse/LIME-2053)

### Other considerations

[LIME-2053]: https://govukverify.atlassian.net/browse/LIME-2053?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: acceptance-tests/run-local-tests.sh [hunk 1]
```diff
@@ -13,6 +13,12 @@ then
 fi
 
 ##### Ask fundamental test questions
+read -p "What is your AWS stack name? [previous=$AWS_STACK_NAME] " AWS_STACK_NAME_NEW
+
+read -p "What is your AWS profile? [previous=$AWS_PROFILE] " AWS_PROFILE_NEW
```
Comment (SarahHillGDS) [inline acceptance-tests/run-local-tests.sh]: The AWS_PROFILE value saved in test-args.conf is not used to authenticate automatically. To run locally we will need to run aws-vault exec dl-dev -- ./run-local-tests.sh first. 
Or could update the test script to detect whether AWS_STACK_NAME is set and conditionally wrap the gradle invocation with aws-vault command.

File: acceptance-tests/src/test/resources/features/UKDrivingLicence/DVA/DVADrivingLicence.feature [hunk 2]
```diff
@@ -13,14 +13,17 @@ Feature: DVA Driving Licence Test
     And I check the page title is Enter your details exactly as they appear on your UK driving licence – GOV.UK One Login
     And I see a form requesting DVA LicenceNumber
 
-  @smoke-build @stub @uat @traffic
+  @smoke-build @stub @uat @traffic @cloudwatch_validation
```
Comment (SarahHillGDS) [inline acceptance-tests/src/test/resources/features/UKDrivingLicence/DVA/DVADrivingLicence.feature]: I noticed that running @cloudwatch_validation in isolation fails. The tag triggers the log assertions but without the full browser journey having run first, vcTxn and feTestStartTime are both null so the step returns an empty result. 
It works fine when the complete @ stub journey runs as part of the same suite, which sets those values as part of navigating to the VC response. The @cloudwatch_validation tag needs to include enough of the journey to populate the correlators, or these assertions should rely solely on the self-contained API scenarios.

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/CloudWatchLogService.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    private List<String> queryByJourneyId(
+            String logGroupName, String journeyId, Instant startTime) {
+        return filterEvents(logGroupName, journeyId, startTime);
+    }
+
+    private List<String> timeWindowQuery(String logGroupName, Instant startTime) {
+        return logsClient
+                .filterLogEvents(
```
Comment (SarahHillGDS) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/CloudWatchLogService.java]: filterLogEvents results are not paginated. This is fine for our logs because of the limited queries but one to be aware of if timeWindowQuery ever ends up on a busy group.

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/step_definitions/CloudWatchLogStepDefs.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+public class CloudWatchLogStepDefs {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CloudWatchLogStepDefs.class);
+
+    private final CloudWatchLogService cloudWatchLogService = new CloudWatchLogService();
+
+    private final String stackName = System.getenv("AWS_STACK_NAME");
+
+    @Then("the {string} lambda logs should contain {string}")
+    public void lambdaLogsShouldContain(String logGroupLogicalId, String expectedMessage) {
```
Comment (SarahHillGDS) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/step_definitions/CloudWatchLogStepDefs.java]: Post POC, log retrieval logic could be de-duplicated

**Review by SarahHillGDS — COMMENTED** [review]
Good handling of journey ID correlation to cover the different journey types. The test suite level PII scan keeps CloudWatchLogService, step definitions and PiiTermsCollector responsibilities clear.

Agree with comment on ticket that Post-POC the main things to pick up would be:
Whether message strings should live behind step definitions
What are appropriate checks for error scenarios where there is no VC

File: acceptance-tests/run-local-tests.sh (hunk 1 above)
Comment (ChrisBates1) [reply acceptance-tests/run-local-tests.sh]: Thanks Sarah, I have removed the AWS_PROFILE prompt, and updated the AWS_STACK_NAME prompt to make it clear its a optional step

If the AWS_STACK_NAME is blank now it should just run as before without any need for AWS auth

File: acceptance-tests/src/test/resources/features/UKDrivingLicence/DVA/DVADrivingLicence.feature (hunk 2 above)
Comment (ChrisBates1) [reply acceptance-tests/src/test/resources/features/UKDrivingLicence/DVA/DVADrivingLicence.feature]: Thanks, I have now removed these steps and tags from the PR. And left this as only PiiScan implementation.

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/CloudWatchLogService.java (hunk 3 above)
Comment (ChrisBates1) [reply acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/CloudWatchLogService.java]: This has now been removed as the cloudwatchLogChecks will be implemented on a different PR. Will address this comment there if required

File: acceptance-tests/run-local-tests.sh [hunk 5]
```diff
@@ -13,6 +13,10 @@ then
 fi
 
 ##### Ask fundamental test questions
+read -p "What is your AWS stack name? (leave blank to skip PII scan) [previous=$AWS_STACK_NAME] " AWS_STACK_NAME_NEW
+
+read -p "What is the ECS log group name for PII scanning? (leave blank to skip) [previous=$ECS_LOG_GROUP_NAME] " ECS_LOG_GROUP_NAME_NEW
```
Comment (SarahHillGDS) [inline acceptance-tests/run-local-tests.sh:18]: When I run local-tests either with an empty ECS_LOG_GROUP_NAME or without the argument altogether the build fails.  The script is still adding the  ECS_LOG_GROUP_NAME to the test-args and empty string is not null. 
Adding an empty string check alongside null check resolves this. Up to you if it makes more sense to keep as empty field or  stop it being created each time in test-args

File: .github/workflows/run-pre-merge-integration-tests.yml [hunk 6]
```diff
@@ -162,6 +162,12 @@ jobs:
           export API_GATEWAY_ID_PRIVATE="${API_GATEWAY_ID_PRIVATE}"
           export API_GATEWAY_ID_PUBLIC="${API_GATEWAY_ID_PUBLIC}"
           export API_GATEWAY_KEY="${APIGW_API_KEY}"
+          export AWS_STACK_NAME="${STACK_NAME}"
+          ECS_LOG_GROUP_SSM=$(aws ssm get-parameter --name "/tests/${STACK_NAME}/ECS_LOG_GROUP_NAME" --region eu-west-2 2>/dev/null || true)
```
Comment (gds-jbentham) [inline .github/workflows/run-pre-merge-integration-tests.yml:166]: Should ECS log checking not be part of the frontend pipeline to avoid us further coupling back and front?

File: acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/CloudWatchLogService.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                .stream()
+                .map(FilteredLogEvent::message)
+                .toList();
+    }
+
+    private String resolveLogGroupName(String stackName, String logGroupLogicalId) {
+        LOGGER.info(
+                "Resolving log group for stack '{}' logical ID '{}'", stackName, logGroupLogicalId);
+        return cfnClient
+                .describeStackResource(
```
Comment (gds-jbentham) [inline acceptance-tests/src/test/java/gov/di_ipv_drivingpermit/utilities/CloudWatchLogService.java:55]: Doesnt need to be one on this PR but would be worth considering if this could be list resources rather than describe and if we could do it once to memory then grab the info we need rather than having to reach back to AWS for each log group

File: acceptance-tests/run-local-tests.sh [hunk 8]
```diff
@@ -101,9 +113,15 @@ echo "BACKEND=${BACKEND}" >> test-args.conf
...
 echo "TAG=${TAG}" >> test-args.conf
+echo "AWS_STACK_NAME=${AWS_STACK_NAME}" >> test-args.conf
+echo "ECS_LOG_GROUP_NAME=${ECS_LOG_GROUP_NAME}" >> test-args.conf
 
 
 ###### Run tests
 actestPath="${PWD}"
-cd ../ && AC_TEST_ONLY=true ./gradlew clean cucumber -P tags="${TAG}"
+if [ -n "${AWS_STACK_NAME}" ]; then
+  cd ../ && aws-vault exec "${AWS_PROFILE:-dl-dev}" -- bash -c "AC_TEST_ONLY=true ./gradlew clean cucumber -P tags='${TAG}'"
```
Comment (gds-jbentham) [inline acceptance-tests/run-local-tests.sh:123]: Can we remove this condition. With the updates my concern is someone sees run local scripts assumes safety but still has an env var like AWS_PROFILE=dl-prod set and while not harmful would start querying and potentially logging pii from prod to their machine.

Without this you would need to do that deliberately. I think our roles would stop this happening anyway but I would rather not leave the option

File: acceptance-tests/run-tests.sh [hunk 9]
```diff
@@ -28,6 +28,8 @@ else
   export ENVIRONMENT="${ENVIRONMENT}"
 fi
 
+export AWS_STACK_NAME="${STACK_NAME}"
```
Comment (gds-jbentham) [inline acceptance-tests/run-tests.sh:31]: Do we need this can we not just use STACK_NAME

File: acceptance-tests/run-tests.sh [hunk 10]
```diff
@@ -66,6 +68,12 @@ if [ "${STACK_NAME}" != "local" ]; then
 
     eval $(echo "export ${NAME}=${VALUE}")
   done
+
+  ECS_LOG_GROUP_SSM=$(aws ssm get-parameter --name "/tests/$STACK_NAME/ECS_LOG_GROUP_NAME" --region eu-west-2 2>/dev/null || true)
```
Comment (gds-jbentham) [inline acceptance-tests/run-tests.sh:72]: Rather than add another param can we not infer ecs name its pretty uniform in naming and has no random characters
